src/gpio/wait_asynch.rs

```diff
@@ -0,0 +1,110 @@
+use core::{
+    future::Future,
+    sync::atomic::{AtomicBool, Ordering},
+    task::Poll,
+};
+extern crate alloc;
+use crate::gpio::{InputMode, InterruptType, Pin, PinDriver};
+use alloc::sync::Arc;
+use atomic_waker::AtomicWaker;
+use esp_idf_sys::{esp_nofail, gpio_intr_disable, EspError};
+
+impl<T: Pin, MODE: InputMode> embedded_hal_async::digital::Wait for PinDriver<'_, T, MODE> {
+    async fn wait_for_high(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::HighLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_low(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::LowLevel)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_rising_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::PosEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_falling_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::NegEdge)?.await;
+        Ok(())
+    }
+
+    async fn wait_for_any_edge(&mut self) -> Result<(), Self::Error> {
+        InputFuture::new(self, InterruptType::AnyEdge)?.await;
+        Ok(())
+    }
+}
+
+pub(crate) struct InputFuture<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode> {
+    // Unfortunately, the Wait trait uses functions that are given a mutable
+    // reference to the pin driver with no explicit lifetime parameter.  Since
+    // the reference has a different lifetime than the struct, we must make this
+    // explicit, or else the borrow checker will believe the reference must live
+    // forever.
+    driver: &'driver_ref mut PinDriver<'driver_struct, T, MODE>,
+    processed_interrupt: Arc<AtomicBool>,
+    atomic_waker: Arc<AtomicWaker>,
+}
+
+impl<'driver_ref, 'driver_struct, T: Pin, MODE: InputMode>
+    InputFuture<'driver_ref, 'driver_struct, T, MODE>
+{
+    pub(crate) fn new(
+        driver: &'driver_ref mut PinDriver<'driver_struct, T, MODE>,
+        interrupt_type: InterruptType,
+    ) -> Result<Self, EspError> {
+        driver.unsubscribe()?;
+        driver.disable_interrupt()?;
+        driver.set_interrupt_type(interrupt_type)?;
+
+        let res = Self {
+            driver,
+            processed_interrupt: Arc::new(AtomicBool::new(false)),
```

Comment:
I would be happier if we don't depend on `alloc` (`Arc`) here. As @MabezDev mentioned, we have a fixed number of GPIO pins, so why not a static array of wakers? Or better yet and as per my previous comment - `NotificationS`?

Reply:
I'll move to notifications